### ISSUE TITLE
MaxTokenBucketizer does not guarentee `max_token_count` holds if padding is used

### ISSUE THREAD

**Issue by nikvaessen** [pytorch/data#788]
### 🐛 Describe the bug

The  `MaxTokenBucketizer` data pipe should probably be aware whether users will pad the resulting list of data to equal length.  As the description gives an example in the audio domain, where padding is frequently used to create a batch of equal length sequences, users will probably run into edge cases where the current behavior will create OOM errors. 

For example, assume a dataset of 5 sequences, with length  [100, 10, 10, 10, 100], and the requirement that no batch has more than 150 tokens. 

The desired behavior would be be 3 batches: [100], [10, 10, 10], and [100]. The current behavior leads to the following:

```python3
import torch as t

from torchdata.datapipes.iter import (
    MaxTokenBucketizer,
    IterableWrapper,
    Mapper,
)


class Sample:
    def __init__(self, length: int):
        self.seq = t.ones((length,))

    def __len__(self):
        return self.seq.shape[0]

    def __lt__(self, other):
        len(self).__lt__(len(other))


def sample_to_tensor(bucket):
    return [x.seq for x in bucket]


def pad_bucket(bucket):
    return t.nn.utils.rnn.pad_sequence(bucket, batch_first=True)


data_lengths = [100, 10, 10, 10, 100]
data = [Sample(x) for x in data_lengths]

dp = IterableWrapper(data)
dp = MaxTokenBucketizer(dp, max_token_count=150)
dp = Mapper(dp, sample_to_tensor)
dp = Mapper(dp, pad_bucket)

for x in dp:
    print(x.shape, f"num_tokens:{x.shape[0] * x.shape[1]}")
```
prints:
```
torch.Size([4, 100]) num_tokens:400
torch.Size([1, 100]) num_tokens:100
```

### Versions

PyTorch version: 1.12.1+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 12.2.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.36

Python version: 3.10.0 (default, Sep 21 2022, 22:36:10) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-5.19.10-arch1-1-x86_64-with-glibc2.36
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3070
Nvidia driver version: 515.76
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.3
[pip3] pytorch-lightning==1.7.7
[pip3] torch==1.12.1+cu113
[pip3] torchaudio==0.12.1+cu113
[pip3] torchdata==0.4.1
[pip3] torchmetrics==0.9.3
[pip3] torchvision==0.13.1+cu113
[conda] Could not collect

**Comment by ejguan** [comment]
I am a little confused here. When input is `[100, 10, 10, 10, 100]` with `max_token_count=150`, I would expect the result is `[100, 10, 10, 10]` and `[100]`. If you want to make batches as `[100], [10, 10, 10], and [100]`, why not to set `max_token_count=100`.

And, could you please clarify why OOM would happen without your proposed `padded_tokens`?

**Comment by nikvaessen** [comment]
I can sketch the following scenario:

The librispeech dataset has audio files with lengths varying from 1 second (16_000 tokens) to 30 seconds (480_000 tokens), with a mean of 12 seconds. When creating batches of training data, each audio sequence needs to be padded to the longest sequence in the batch. We can use the `MaxTokenBucketizer` to create batches with at most `x` tokens. When reproducing fine-tuning with wav2vec2, for example, the value is [3_200_000](https://github.com/facebookresearch/fairseq/blob/699ab190147e3afff0201c956d8145d158bc6f28/examples/wav2vec/config/finetuning/base_960h.yaml#L20) (200 seconds per batch). It is used directly as batch size value, as well as controlling that batches stay small enough so you don't exceed GPU memory. 

Note that you can only pad after you know your batch, as you need to know the longest sequence. If we use the current implementation of `MaxTokenBucketizer`, it's possible that batches exceed the 200 second batch size. For example, a buffer with 10 files of 12 seconds, and 5 files of 15 seconds, will create a batch of 195 seconds, but after padding it will be 225 seconds. This can be harmful for reproduction effort and for potentially result in OOM.

**Comment by ejguan** [comment]
I see. Then, a `padden_tokens` is not enough to support this feature. An option to specify the total padded length for a batch would be a better option. And, this number seems always larger than `max_token_count`. It looks more reasonable to me to add another DataPipe for your specific use case rather than using the `MaxTokenBucketizer`. WDYT?

**Comment by nikvaessen** [comment]
Do you mean that the padding should be done within (a new version of) `MaxTokenBucketizer`? I think it is reasonable to have the resulting output (with `padden_tokens=True` if your use-case desires it) be passed through `Collator` to create a batch of the output list.

**Comment by ejguan** [comment]
NVM. Let's add this argument to the `DataPipe`. I will review your PR shortly.

**Comment by ejguan** [comment]
Another question. Does this feature request also needs data will be batched starting from shorter data to longer data. Let's say you have [100, 10, 10, 10, 100] as input with `max_token_count`. The actual result will be `[10, 10, 10], [100], [100]`. This would guarantee that data with closer length within the buffer will be aggregated into a batch.

**Comment by nikvaessen** [comment]
I think ideally, you would solve a knapsack-like problem to create a batch with largest length from the items in the buffer (and thus get the batch with the least amount of padding possible). But I'm not sure how much that actually matters in practice, and whether the extra complexity would be worth it. 

I think the current behavior, with taking the smallest samples from the buffer, is fine.

**Comment by nikvaessen** [comment]
Although, perhaps it might be worth it to consider randomly filling the batch instead of using the smallest tokens, to increase the randomness of the generated batches. Or instead of starting at index 0 of the heap (the smallest index), take a random index `i`, between 0 and len(buffer), and fill the batch with samples `i, i+1, ...` until the batch would exceed `max_token_count`. If `i` reaches `len(buffer)`, we can start taking samples `i-1, ...`.

**Comment by ejguan** [comment]
IMHO, taking the smallest is equivalent to taking the largest from the buffer in terms of minimizing padding.

> Although, perhaps it might be worth it to consider randomly filling the batch instead of using the smallest tokens, to increase the randomness of the generated batches. Or instead of starting at index 0 of the heap (the smallest length), take a random index `i`, between `0` and `len(buffer)`, and fill the batch with samples `i, i+1, ...` until the batch would exceed `max_token_count`. If `i` reaches `len(buffer)`, we can start taking samples `i-1, ...`. Or take samples in the order `i+1, i-1, i+2, i-2, ...`.

I think it makes this DataPipe too complicated. Users should be able to achieve the similar result by using a smaller `buffer_size` and doing `shuffle` before and after this operation to achieve the similar result.

**Comment by nikvaessen** [comment]
> I think it makes this DataPipe too complicated. Users should be able to achieve the similar result by using a smaller buffer_size and doing shuffle before and after this operation to achieve the similar result.

I agree. But perhaps worth to document this in the PR.
